src/link/SpirV.zig

```diff
@@ -139,55 +118,69 @@ pub fn flushModule(self: *SpirV, comp: *Compilation) !void {
     const module = self.base.options.module.?;
     const target = comp.getTarget();
 
+    var spirv_module = codegen.SPIRVModule.init(target, self.base.allocator);
+    defer spirv_module.deinit();
+
+    // Allocate an ID for every declaration before generating code,
+    // so that we can access them before processing them.
+    // TODO: We're allocating an ID unconditionally now, are there
+    // declarations which don't generate a result?
```

Comment:
I believe this question is answered by asking `Type.hasCodeGenBits()` for the Type of the Decl. The frontend code does not call `updateDecl` when false, but since you're iterating over `module.decl_table` manually, you'll have to do this check.

Note also that in my branch #8554, the global decl_table is gone, and I modified the C and SPIR-V backends to keep track of a set of decls in response to updateDecl and freeDecl, like this:

```diff
diff --git a/src/link/SpirV.zig b/src/link/SpirV.zig
index 7a35752f6..0a7fec941 100644
--- a/src/link/SpirV.zig
+++ b/src/link/SpirV.zig
@@ -37,9 +37,14 @@ pub const FnData = struct {
 
 base: link.File,
 
-// TODO: Does this file need to support multiple independent modules?
+/// TODO: Does this file need to support multiple independent modules?
 spirv_module: codegen.SPIRVModule,
 
+/// This linker backend does not try to incrementally link output SPIR-V code.
+/// Instead, it tracks all declarations in this table, and iterates over it
+/// in the flush function.
+decl_table: std.AutoArrayHashMapUnmanaged(*Module.Decl, void) = .{},
+
 pub fn createEmpty(gpa: *Allocator, options: link.Options) !*SpirV {
     const spirv = try gpa.create(SpirV);
     spirv.* = .{
@@ -88,6 +93,7 @@ pub fn openPath(allocator: *Allocator, sub_path: []const u8, options: link.Optio
 }
 
 pub fn deinit(self: *SpirV) void {
+    self.decl_table.deinit(self.base.allocator);
     self.spirv_module.deinit();
 }
 
@@ -95,6 +101,9 @@ pub fn updateDecl(self: *SpirV, module: *Module, decl: *Module.Decl) !void {
     const tracy = trace(@src());
     defer tracy.end();
 
+    // Keep track of all decls so we can iterate over them on flush().
+    _ = try self.decl_table.getOrPut(self.base.allocator, decl);
+
     const fn_data = &decl.fn_link.spirv;
     if (fn_data.id == null) {
         fn_data.id = self.spirv_module.allocId();
@@ -118,6 +127,7 @@ pub fn updateDeclExports(
 ) !void {}
 
 pub fn freeDecl(self: *SpirV, decl: *Module.Decl) void {
+    self.decl_table.removeAssertDiscard(decl);
     var fn_data = decl.fn_link.spirv;
     fn_data.code.deinit(self.base.allocator);
     if (fn_data.id) |id| self.spirv_module.freeId(id);
@@ -164,19 +174,15 @@ pub fn flushModule(self: *SpirV, comp: *Compilation) !void {
     defer all_buffers.deinit();
 
     // Pre-allocate enough for the binary info + all functions
-    try all_buffers.ensureCapacity(module.decl_table.count() + 1);
+    try all_buffers.ensureCapacity(self.decl_table.count() + 1);
 
     all_buffers.appendAssumeCapacity(wordsToIovConst(binary.items));
 
-    for (module.decl_table.items()) |entry| {
-        const decl = entry.value;
-        switch (decl.typed_value) {
-            .most_recent => |tvm| {
-                const fn_data = &decl.fn_link.spirv;
-                all_buffers.appendAssumeCapacity(wordsToIovConst(fn_data.code.items));
-            },
-            .never_succeeded => continue,
-        }
+    for (self.decl_table.items()) |entry| {
+        const decl = entry.key;
+        if (!decl.has_tv) continue;
+        const fn_data = &decl.fn_link.spirv;
+        all_buffers.appendAssumeCapacity(wordsToIovConst(fn_data.code.items));
     }
 
     var file_size: u64 = 0;
```

so one of us is going to have to deal with this conflict :)

Reply:
I previously had a look at your branch to see the difference. The change didn't look like much work, so if you think your branch is soon ready for merging i'm also with waiting a bit and then fix the conflict before merging this PR.

Reply:
I wouldn't mind resolving the conflict either - I think this can be merged as soon as the CI is green (except lavahut which is broken at the moment)